File: src/com/example/javas/medium/M005_TaskScheduler.java
/*
SYSTEM DESIGN PROBLEM:
Design a simple Task Scheduler (Simplified for 25 min interview)

REQUIREMENTS:
- Add tasks with priority (1=highest, 5=lowest)
- Execute next task (highest priority first, FIFO if same priority)
- Get pending task count

CONSTRAINTS:
- Priority range: 1-5
- Simple in-memory storage
- Tasks are just strings with priority

DESIGN APPROACH:
1. Use PriorityQueue to store tasks by priority
2. Custom Task class with priority and order tracking
3. Counter for insertion order (to maintain FIFO for same priority)

DATA STRUCTURES USED:
- PriorityQueue<Task>: Auto-sorts by priority then insertion order
- Task class: id, name, priority, insertionOrder

TIME COMPLEXITY:
- addTask(): O(log n)
- executeNext(): O(log n)
- getPendingCount(): O(1)

SPACE COMPLEXITY: O(n) where n is number of pending tasks
*/

package com.example.javas.medium;

import java.util.*;

public class M005_TaskScheduler {
    
    static class Task {
        String name;
        int priority;
        int insertionOrder;
        
        public Task(String name, int priority, int insertionOrder) {
            // DONE: Initialize task fields
            this.name = name;
            this.priority = priority;
            this.insertionOrder = insertionOrder;
        }
    }
    
    static class TaskScheduler {
        private PriorityQueue<Task> taskQueue;
        private int counter;
        
        public TaskScheduler() {
            // DONE: Initialize PriorityQueue with custom comparator
            // Comparator logic:
            // 1. Compare by priority (lower number = higher priority)
            // 2. If same priority, compare by insertionOrder (FIFO)
            this.counter = 0;
            this.taskQueue = new PriorityQueue<>((t1, t2) -> {
                if (t1.priority != t2.priority) {
                    return t1.priority - t2.priority; // Lower priority number = higher priority
                }
                return t1.insertionOrder - t2.insertionOrder; // FIFO for same priority
            });
        }
        
        public void addTask(String taskName, int priority) {
            // DONE: Implement add task
            // 1. Create new Task with counter as insertionOrder
            // 2. Add to taskQueue
            // 3. Increment counter
            Task task = new Task(taskName, priority, counter);
            taskQueue.offer(task);
            counter++;
        }
        
        public String executeNext() {
            // DONE: Implement execute next task
            // 1. Poll from taskQueue (gets highest priority)
            // 2. Return task name or "No tasks" if empty
            Task task = taskQueue.poll();
            return task != null ? task.name : "No tasks";
        }
        
        public int getPendingCount() {
            // DONE: Return number of pending tasks in queue
            return taskQueue.size();
        }
    }
    
    // TEST CASES
    public static void main(String[] args) {
        TaskScheduler scheduler = new TaskScheduler();
        
        scheduler.addTask("Email Report", 3);
        scheduler.addTask("Fix Bug", 1);
        scheduler.addTask("Write Docs", 3);
        scheduler.addTask("Deploy App", 1);
        scheduler.addTask("Code Review", 2);
        
        System.out.println("Pending tasks: " + scheduler.getPendingCount());
        System.out.println();
        
        System.out.println("Executing tasks:");
        while (scheduler.getPendingCount() > 0) {
            System.out.println("- " + scheduler.executeNext());
        }
        System.out.println();
        
        System.out.println("Execute when empty: " + scheduler.executeNext());
        System.out.println("Pending tasks: " + scheduler.getPendingCount());
    }
}

File: pythonX/medium/m005_task_scheduler.py
"""
M005: Design Task Scheduler

Summary:
Design a task scheduler with priority queue.

Requirements:
Implement the TaskScheduler class:
- add_task(task_name, priority): Adds task with priority (1=highest, 5=lowest)
- execute_next(): Executes and returns next highest priority task
- get_pending_count(): Returns pending task count

Approach:
Use heap (priority queue) to manage tasks by priority and insertion order.

Data Structures:
- heap: min-heap of (priority, insertion_order, task_name)
- int: counter for insertion order

Big O Analysis:
Time Complexity: O(log n) for add_task and execute_next
Space Complexity: O(n) where n is pending tasks
"""

import heapq


class TaskScheduler:
    
    def __init__(self):
        """
        TODO: Initialize heap and counter
        """
        pass
    
    def add_task(self, task_name: str, priority: int):
        """
        TODO: Add task to heap
        1. Push (priority, counter, task_name) to heap
        2. Increment counter
        """
        pass
    
    def execute_next(self) -> str:
        """
        TODO: Execute next task
        1. Pop from heap
        2. Return task name or "No tasks"
        """
        pass
    
    def get_pending_count(self) -> int:
        """
        TODO: Return heap size
        """
        pass


if __name__ == "__main__":
    # Test Cases
    
    print("Test 1: Priority scheduling")
    scheduler = TaskScheduler()
    
    scheduler.add_task("Email Report", 3)
    scheduler.add_task("Fix Bug", 1)
    scheduler.add_task("Write Docs", 3)
    scheduler.add_task("Deploy App", 1)
    scheduler.add_task("Code Review", 2)
    
    print(f"Pending tasks: {scheduler.get_pending_count()}")  # Expected: 5
    
    print("\nExecuting tasks:")
    while scheduler.get_pending_count() > 0:
        print(f"- {scheduler.execute_next()}")
    # Expected order: Fix Bug, Deploy App, Code Review, Email Report, Write Docs
    
    print(f"\nExecute when empty: {scheduler.execute_next()}")  # Expected: No tasks
    print(f"Pending tasks: {scheduler.get_pending_count()}")    # Expected: 0
